allow changing rulesets through CLI: osu set-ruleset shortname

WebFreak001 opened the pull request:
e.g.:
osu://ruleset/osu
osu://ruleset/taiko
osu://ruleset/mania
osu://ruleset/fruits
osu://ruleset/pippidon

I just personally want this for my arcade cab, which allows switching games or gamemodes by plugging in/out controllers via USB, where I want to switch to my custom ruleset (https://github.com/WebFreak001/katsudon) when 2 taiko controllers are plugged in and switch to regular taiko when just a single controller is plugged in.

peppy commented:
I'm not sure about this. I don't think we want url schema to allow mutating game state. It seems like an abuse vector.

WebFreak001 commented:
in that case I will change it to instead be a new IPC command, e.g.: `./osu set-ruleset shortname`

WebFreak001 commented:
I changed this to a new IPC packet now, where the API is intended to be also used with other similar setter-like calls or calls that are just taking in some simple string arguments. (e.g. there could be interactive settings changes through CLI with this)

Note that I also adjusted the existing test here, however I was unable to actually test it since that test was broken on my machine prior to the PR already / it wasn't as simple as just running the test project.

WebFreak001 committed:
add new IPC to change rulesets

e.g.:
osu set-ruleset osu
osu set-ruleset taiko
osu set-ruleset mania
osu set-ruleset fruits
osu set-ruleset pippidon

I just personally want this for my arcade cab, which allows switching games or gamemodes by plugging in/out controllers via USB, where I want to switch to my custom ruleset (https://github.com/WebFreak001/katsudon) when 2 taiko controllers are plugged in and switch to regular taiko when just a single controller is plugged in.

WebFreak001 commented:
note: CodeFactor fail is due to code having the same structure as the other IPC code, which all uses 2 classes in one file right now

WebFreak001 commented:
> Heck use your custom build for the arcade cab or something. Which I guess you probably don't want to do since it won't submit anything.

this is what I'm doing already and is no problem for me since I don't actually want to submit anything. (it's running mostly offline actually) I'm using this to also perform other tasks such as changing the local guest account to different ones. (guest account with custom name)

I just thought this is probably a feature more people want to have (e.g. changing ruleset, in-game settings such as skin, slider speed, etc. via scripts) - a janky script wouldn't work with most features I'm using this for, but didn't PR yet.

> Explicitly opt-in with giant warning bells that if you enable this then you allow stuff to happen to your game that possibly you did not do.

First, I think the risk here is not that high, a malicious actor would require permissions to spawn processes or talk with a named pipe / unix socket at an exact file location to use this API - not things, that you would ever be able in untrusted scenarios. In that case any malicious actor most likely already is running in their own process, where they could e.g. most likely write to the internal osu database/config files to do more serious stuff messing with the game. With future extended possibilities this might become more needed though.

This idea actually also came from a prior form of the state broadcast PR, since this is a lot easier to integrate in the existing code, is a lot less work for the consumer than spinning up a websocket client, is much easier for the API users not caring about receiving any state and won't require so much extra state handling in the osu side as well as being definitely locked to the local machine, avoiding security issues with sockets.

Adding it to the websockets is definitely not a good idea without first having some sort of authentication (e.g. confirmation notification "browser/app xyz wants to control osu" + management settings), but if we do tackle this work this could be the best integration UX-wise, allowing fine user control, allowing even random websites to change settings and do things in-game (I can see a lot of use-cases here).

If you want I can instead work on the security scheme and integration in osu for how this could be integrated with the future websocket integration - it's a lot larger in scope and code complexity, but would offer even cooler things I did not think about initially with this PR yet.

peppy commented:
For now I think it's best you build around this idea in a fork. I definitely want to see an arcade mode added, but alone, this single API request feels hard to add in terms of added complexity for minimal gains.

If we're going to add a remote control API like this, I'd hope it's more fleshed out and serves more than one person's usage.

Going to close this out, but please consider maintaining a fork/branch with your arcade work and looping me in if you have cool things to show 😄
